bionemo-recipes/recipes/esm2_native_te/dataset.py

```diff
@@ -120,14 +128,49 @@ def tokenize_function(examples):
             seed=seed,
         )
 
+    if use_token_packing:
+        assert use_sequence_packing, "Token packing assumes sequence packing."
+        assert isinstance(tokenized_dataset, datasets.IterableDataset), "Token packing requires a streaming dataset."
+        tokenized_dataset = TokenPackingDataset(tokenized_dataset, max_length=micro_batch_size)
+
     train_dataloader = StatefulDataLoader(
         tokenized_dataset,
         sampler=sampler,
-        batch_size=micro_batch_size,
+        # When we're using token packing, we don't need the dataloader to handle the batch size.
+        batch_size=micro_batch_size if not use_token_packing else None,
         collate_fn=data_collator,
         num_workers=num_workers,
         pin_memory=True,
         persistent_workers=True,
     )
 
     return train_dataloader, dataset if sampler is None else sampler
+
+
+@dataclass
+class TokenPackingDataset(torch.utils.data.IterableDataset):
+    """Dataset that uses sequence packing to construct batches with variable length up to a maximum number of tokens."""
+
+    dataset: datasets.IterableDataset
+    max_length: int
```

Comment:
so max length here is the maximum number of tokens per batch? Eg if you were thinking about `BSHD` with no padding it would be `B * S`?

Reply:
If so, `max_tokens_per_batch` is probably overly verbose but more clear. `max_tokens` maybe?

Reply:
So... here is `max_length` like `max_sequence_length`? I think I was just confused thinking about the naming and the T parameter.